File: PARAGENESIS/best.cs

/*
 *  file:	best.c
 *
 *  		Copyright (c) 1986, 1990 by John J. Grefenstette
 *
 *  purpose:	input, maintenance and output of best structures
 *
 *  modified:	15 apr 86
 *		12 nov 86: pass Length to Pack() and Unpack()
 *		16 dec 86: don't print Bestsize to minfile in Printbest()
 *		15 sep 90: handle float representation in i/o
 */

#include "Pextern.h"

static double worst_value;	/* worst value in the Bestset		*/
static int worst;		/* pointer to element with worst_value	*/


double BEST(ParallelNew)
STRUCTURE:physical ParallelNew;
{
  with(physical)
    where (Index < Popsize)
      if (Maxflag)
	return (>?= ParallelNew.Perf);
      else
	return (<?= ParallelNew.Perf);
}

PSavebest(ParallelNew)
STRUCTURE:physical ParallelNew;
{
  register int j;
  register int k;
  int found,curr_index;
  int:physical untested;
  
  with(physical) {
    untested = 1; 
    
    where(Index < Popsize) {
      if (Bestsize < Savesize) {
	
	while((Bestsize < Savesize) && (+= untested))
	  where(untested) {
	    where(ParallelNew.Perf == BEST(ParallelNew)) {
	      curr_index = (int)Index;
	      where(curr_index == Index) {
		untested = 0;
		for (j=0, found=0; j<Bestsize && (!found); j++)
		  for (k=0, found=1; (k<Bytes) && (found); k++)
		    found = (int)(ParallelNew.Gene[k] == Bestset[j].Gene[k]);
		if (!found) {
		  for (k=0; k<Bytes; k++)
		    Bestset[Bestsize].Gene[k] = (char)ParallelNew.Gene[k];
		  Bestset[Bestsize].Perf = (double)ParallelNew.Perf;
		  Bestset[Bestsize].Gen = Gen;
		  Bestset[Bestsize].Trials = Trials;
		  Bestsize++;
		}
	      }
	    }
	  }
	
	if (Bestsize == Savesize)
	  {
	    /* find worst element in Bestset */
	    worst_value = Bestset[0].Perf;
	    worst = 0;
	    for (j=1; j<Savesize; j++)
	      {
		if (BETTER(worst_value,Bestset[j].Perf))
		  {
		    worst_value = Bestset[j].Perf;
		    worst = j;
		  }
	      }
	  }
      }
      else
	{
	  if (BETTER(BEST(ParallelNew), worst_value)) {
	    
	    while (+= untested)
	      where(untested) {
		if (BETTER(worst_value, BEST(ParallelNew)))
		  untested = 0;
		else 
		  where(ParallelNew.Perf == BEST(ParallelNew)) {
		    curr_index = (int)Index;
		    where(curr_index == Index) {
		      untested = 0;
		      for (j=0, found=0; j<Bestsize && (!found); j++)
			for (k=0, found=1; (k<Bytes) && (found); k++)
			  found = (int)(ParallelNew.Gene[k] == 
					Bestset[j].Gene[k]);
		      if (!found) {
			for (k=0; k<Bytes; k++)
			  Bestset[worst].Gene[k] = (char)ParallelNew.Gene[k];
			Bestset[worst].Perf = (double)ParallelNew.Perf;
			Bestset[worst].Gen = Gen;
			Bestset[worst].Trials = Trials;
			
			worst_value = Bestset[0].Perf;
			worst = 0;
			for (j=1; j<Savesize; j++)
			  if (BETTER(worst_value, Bestset[j].Perf)) {
			    worst_value = Bestset[j].Perf;
			    worst = j;
			  }
		      }
		    }
		  }
	      }
	  }
	}
    }
  }
}

Printbest()
{
  /*	Write the Best structures out to the Bestfile.	*/
  
  register int i;
  register int j;
  FILE *fp, *fopen();
  
  Trace("Printbest entered");
  Time(0,"");
  
  fp = fopen(Bestfile, "w");
  /* fprintf(fp, "%d\n", Bestsize); */
  for (i=0; i<Bestsize; i++)
    {
      Unpack(Bestset[i].Gene, Bitstring, Length);
      if (Floatflag)
	{
	  FloatRep(Bitstring, Vector, Genes);
	  for (j=0; j<Genes; j++)
	    {
	      fprintf(fp, Gene[j].format, Vector[j]);
	    }
	}
      else
	{
	  for (j=0; j<Length; j++)
	    {
	      if ((j > 0) && ((j % 8) == 0)) fprintf(fp, " ");
	      fprintf(fp, "%c", Bitstring[j]);
	    }
	}
      fprintf(fp, "  %11.4e ", Bestset[i].Perf);
      fprintf(fp, " %4d  %4d\n", Bestset[i].Gen, Bestset[i].Trials);
    }
  
  fclose(fp);
  Trace("Printbest completed");
}


Readbest()
{
  /*   Read the Best structures in from the Bestfile    */
  /*   (during a Restart)                               */
  
  int i;
  FILE *fp, *fopen();
  int status;
  
  if (Savesize)
    {
      Trace("Readbest entered");
      
      fp = fopen(Bestfile, "r");
      if (fp == NULL) 
	{
	  Bestsize = 0;
	  return;
	}
      
      Bestsize = 0;
      status = 1;
      if (Floatflag)
	{
	  for (i = 0; i < Genes && status != EOF; i++)
	    {
	      status = fscanf(fp, "%lf", &Vector[i]);		
	    }
	}
      else
	status = fscanf(fp, "%s", Bitstring);
      
      while (status != EOF)
	{
	  if (Floatflag)
	    StringRep(Vector, Bitstring, Genes);
	  
	  Pack(Bitstring, Bestset[Bestsize].Gene, Length);
	  
	  fscanf(fp, "%lf ", &Bestset[Bestsize].Perf);
	  fscanf(fp, "%d  %d", &Bestset[Bestsize].Gen,
		 &Bestset[Bestsize].Trials);
	  Bestsize++;
	  
	  /* get the next structure */
	  if (Floatflag)
	    for (i = 0; i < Genes && status != EOF; i++)
	      status = fscanf(fp, "%lf", &Vector[i]);		
	  else
	    status = fscanf(fp, "%s", Bitstring);
	}
      fclose(fp);
      
      /* find worst element in Bestset */
      worst_value = Bestset[0].Perf;
      worst = 0;
      for (i=1; i<Bestsize; i++)
	{
	  if (BETTER(worst_value, Bestset[i].Perf))
	    {
	      worst_value = Bestset[i].Perf;
	      worst = i;
	    }
	}

      Time(1,"Readbest");
      Trace("Readbest completed");
    }
}

/**** end of file ****/

File: GAucsd/src/best.c
/*************************************************************/
/*                                                           */
/*  Copyright (c) 1986                                       */
/*  John J. Grefenstette                                     */
/*  Navy Center for Applied Research in AI                   */
/*  Naval Research Laboratory                                */
/*                                                           */
/*  Permission is hereby granted to copy all or any part of  */
/*  this program for free distribution.   The author's name  */
/*  and this copyright notice must be included in any copy.  */
/*                                                           */
/*************************************************************/

/*
 *  file:	best.c
 *
 *  author:	John J. Grefenstette
 *
 *  created:	1983
 *
 *  purpose:	input and output of best structures
 *
 *  modified:	15 apr 86
 *		16 dec 86: don't print Bestsize to minfile in Printbest()
 *      02 nov 90: store phenotypes, allow duplicates when -a
 */

#define   EXTERN
#include "global.h"

extern void Pack();
extern void Unpack();
extern double _eval();

static double max;	/* maximum (worst) performance in the Bestset */
static int maxptr;	/* pointer to structure with perf = max */


Savebest(i)
register int i;
{
	/*  Save the ith structure in current population  */
	/*  if it is one of the Savesize best seen so far */

	register int j;		/* loop control var */
	register int k;

	int found;
	if (Bestsize < Savesize)
	{
		if (!Allflag)
		{
			/* Bestsize is the number saved so far, so  		*/
			/* there are still empty slots in the Bestset.		*/
			/* Check if an identical structure is already there	*/

			for (j = 0; j < Bestsize; j++)
			{
				for (k = 0, found = 1; (k < Bytes) && found; k++)
					found = (New[i].Gene[k] == Bestset[j].Gene[k]);
				if (found && Bestset[j].Gen > Lastzoom) return;
			}
		}

		/* insert ith structure */
		for (k = 0; k < Bytes; k++)
			Bestset[Bestsize].Gene[k] = New[i].Gene[k];

		/* obtain the phenotype, if available */
		*(Bestset[Bestsize].Phene) = '\0';
		_eval(Bestset[Bestsize].Phene, -Phenesize);

		Bestset[Bestsize].Perf = New[i].Perf;
		Bestset[Bestsize].Gen = Gen;
		Bestset[Bestsize].Trials = Trials;
		Bestsize++;

		if (Bestsize == Savesize)
		{
			/* find worst element in Bestset */
			max = Bestset[0].Perf;
			maxptr = 0;
			for (j = 1; j < Savesize; j++)
			{
				if (max < Bestset[j].Perf)
				{
					max = Bestset[j].Perf;
					maxptr = j;
				}
			}
		}
	}
	else if (New[i].Perf < max) 	/* then save New[i] */
	{
		if (!Allflag)   	/* unless its already there */
		{
			for (j = 0; j < Bestsize; j++)
			{
				for (k = 0, found = 1; (k < Bytes) && found; k++)
					found = (New[i].Gene[k] == Bestset[j].Gene[k]);
				if (found && Bestset[j].Gen > Lastzoom) return;
			}
		}

		/* overwrite the worst one */
		for (k = 0; k < Bytes; k++)
			Bestset[maxptr].Gene[k] = New[i].Gene[k];

		/* obtain the phenotype, if available */
		*(Bestset[maxptr].Phene) = '\0';
		_eval(Bestset[maxptr].Phene, -Phenesize);

		Bestset[maxptr].Perf = New[i].Perf;
		Bestset[maxptr].Gen = Gen;
		Bestset[maxptr].Trials = Trials;

		/* find worst element in Bestset */
		max = Bestset[0].Perf;
		maxptr = 0;
		for (j = 1; j < Savesize; j++)
		{
			if (max < Bestset[j].Perf)
			{
				max = Bestset[j].Perf;
				maxptr = j;
			}
		}
	}
}


Printbest()
{
	/*	Write the Best structures out to the Bestfile.	*/

	register int i;
	register int j;
	register int k;
	register int posn;
	FILE *fp, *fopen();

	Trace("Printbest entered");

	fp = fopen(Bestfile, "w");
	for (i = 0; i < Bestsize; i++)
	{
		Unpack(Bestset[i].Gene, Buff);
		for (j = k = posn = 0; j < Length; j++)
		{
			if (GAgenes)
			{
				if (j == posn)
				{
					if (j) fprintf(fp, " ");
					posn = GAposn[k++];
					if (posn < 0) posn = -posn;
				}
			}
			fprintf(fp, "%1d", Buff[j]);
		}

		fprintf(fp, "  %11.4e ", Bestset[i].Perf);
		fprintf(fp, "%4d %4ld ", Bestset[i].Gen, Bestset[i].Trials);
		fprintf(fp, "%s\n\n",    Bestset[i].Phene);
	}

	fclose(fp);

	Trace("Printbest completed");
}


Readbest()
{
	/*   Read the Best structures in from the Bestfile    */
	/*   (during a Restart)                               */

	int i, j, k;
	char *s;
	FILE *fp, *fopen();
	static char *dummy = ".";

	if (Savesize)
	{
		Trace("Readbest entered");

		fp = fopen(Bestfile, "r");
		if (fp == NULL) 
		{
			Bestsize = 0;
			return;
		}

		for (Bestsize = 0; (fscanf(fp, "%1d", &k) != EOF); Bestsize++)
		{
			Buff[0] = k;
			for (j = 1; j < Length; j++)
			{
				fscanf(fp, "%1d", &k);
				Buff[j] = k;
			}
			Pack(Buff, Bestset[Bestsize].Gene);

			fscanf(fp, " %lf", &Bestset[Bestsize].Perf);
			fscanf(fp, " %d ", &Bestset[Bestsize].Gen);
			fscanf(fp, " %ld", &Bestset[Bestsize].Trials);

			/* read phenotype description, if any */
			s = dummy;
			for (k = -1; k < Phenesize && *s != '\n'; k += strlen(s))
				s = fgets(Bestset[Bestsize].Phene+k, Phenesize+1-k, fp);

			if (k < Phenesize) Bestset[Bestsize].Phene[k-2] = '\0';
			else Error("Phenotype description too long");
		}
		fclose(fp);

		/* find worst element in Bestset */
		max = Bestset[0].Perf;
		maxptr = 0;
		for (i = 1; i < Bestsize; i++)
		{
			if (max < Bestset[i].Perf)
			{
				max = Bestset[i].Perf;
				maxptr = i;
			}
		}

		Trace("Readbest completed");
	}
}

/**** end of file ****/

